hash_reader.go

```diff
@@ -49,56 +49,35 @@ func (hashReader *hashReader) SymbolTable() ion.SymbolTable {
 }
 
 func (hashReader *hashReader) Next() bool {
-	switch hashReader.currentType {
-	case ion.ListType:
-		fallthrough
-	case ion.SexpType:
-		fallthrough
-	case ion.StructType:
-		if hashReader.IsNull() {
-			err := hashReader.hasher.scalar(hashReader)
-			if err != nil {
-				return false
-			}
-		} else {
-			err := hashReader.StepIn()
-			if err != nil {
-				return false
-			}
+	if hashReader.currentType != ion.NoType {
+		if ion.IsContainer(hashReader.currentType) {
+			if hashReader.IsNull() {
+				err := hashReader.hasher.scalar(hashReader)
+				if err != nil {
+					return false
+				}
+			} else {
+				err := hashReader.StepIn()
+				if err != nil {
+					return false
+				}
 
-			err = hashReader.traverse()
-			if err != nil {
-				return false
-			}
+				err = hashReader.traverse()
+				if err != nil {
+					return false
+				}
 
-			err = hashReader.StepOut()
+				err = hashReader.StepOut()
+				if err != nil {
+					return false
+				}
+			}
+		} else if ion.IsScalar(hashReader.currentType) {
+			err := hashReader.hasher.scalar(hashReader)
 			if err != nil {
 				return false
 			}
 		}
```

Comment:
nit: I like the exhaustiveness of the `switch`. The last else-if here is not exhaustive (or rather it relies on the implementations of isScalar and isContainer to separate the space of types completely). If we can update the if statement here (add an else branch for example) that would be better.

Reply:
I just reworked this if-statement from

if container:
    if IsNull:
        scalar logic
    else:
        container logic
else if scalar:
    scalar logic

to

if scalar or IsNull:
    scalar logic
else: 
    container logic

Reply:
Reworked if-statement to be more exhaustive.